test/performance/odp_dma_perf.c

```diff
@@ -0,0 +1,876 @@
+	       "    average transfer speed:       ",
+	       config->num_rounds, avg_time);
+	print_humanised_speed(avg_speed);
+	printf("\n=============================================\n");
+}
+
+static int run_dma_sync(test_config_t *config)
+{
+	odp_dma_transfer_param_t trs_params[config->dma_rounds];
+	uint32_t trs_lengths[config->dma_rounds];
+	odp_time_t start, end;
+	uint32_t num_rounds = config->num_rounds, offset;
+
+	config->test_case_api.trs_base_fn(config, trs_params, trs_lengths);
+	start = odp_time_local_strict();
+
+	while (num_rounds--) {
+		offset = 0U;
+
+		for (int i = 0; i < config->dma_rounds; ++i) {
+			config->test_case_api.trs_dyn_fn(config, offset, trs_lengths[i]);
+
+			if (odp_dma_transfer(config->dma_config.handle, &trs_params[i], NULL)
+			    <= 0) {
+				ODPH_ERR("Error starting a sync DMA transfer.\n");
+				return -1;
+			}
+
+			offset += trs_lengths[i];
+		}
+	}
+
+	end = odp_time_local_strict();
+	print_results(config, odp_time_diff_ns(end, start));
+	return 0;
+}
+
+static int configure_dma_event_completion(test_config_t *config)
+{
+	int ret;
+	odp_dma_pool_param_t pool_param;
+	odp_queue_param_t queue_param;
+
+	ret = odp_schedule_config(NULL);
+
+	if (ret < 0) {
+		ODPH_ERR("Error configuring scheduler.\n");
+		return -1;
+	}
+
+	odp_dma_pool_param_init(&pool_param);
+	pool_param.num = config->num_in_seg;
+	config->dma_config.pool = odp_dma_pool_create("odp_dma_perf_events", &pool_param);
+
+	if (config->dma_config.pool == ODP_POOL_INVALID) {
+		ODPH_ERR("Error creating DMA event completion pool.\n");
+		return -1;
+	}
+
+	odp_queue_param_init(&queue_param);
+	queue_param.type = ODP_QUEUE_TYPE_SCHED;
+	queue_param.sched.sync = ODP_SCHED_SYNC_PARALLEL;
+	queue_param.sched.prio = odp_schedule_default_prio();
+	queue_param.sched.group = ODP_SCHED_GROUP_ALL;
+	config->dma_config.compl_q = odp_queue_create("odp_dma_perf_queue", &queue_param);
+
+	if (config->dma_config.compl_q == ODP_QUEUE_INVALID) {
+		ODPH_ERR("Error creating DMA completion queue.\n");
+		return -1;
+	}
+
+	config->dma_config.compl_ev = odp_dma_compl_alloc(config->dma_config.pool);
+
+	if (config->dma_config.compl_ev == ODP_DMA_COMPL_INVALID) {
+		ODPH_ERR("Error creating DMA completion event.\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+static int configure_dma_completion_params(test_config_t *config,
+					   odp_dma_compl_param_t compl_params[])
+{
+	for (int i = 0; i < config->dma_rounds; ++i) {
+		odp_dma_compl_param_init(&compl_params[i]);
+
+		if (config->compl_modes.modes[i] == ODP_DMA_COMPL_EVENT) {
+			compl_params[i].compl_mode = ODP_DMA_COMPL_EVENT;
+
+		} else {
+			compl_params[i].compl_mode = ODP_DMA_COMPL_POLL;
+			compl_params[i].transfer_id =
+				odp_dma_transfer_id_alloc(config->dma_config.handle);
+
+			if (compl_params[i].transfer_id == ODP_DMA_TRANSFER_ID_INVALID) {
+				ODPH_ERR("Error creating DMA transfer ID.\n");
+				return -1;
+			}
+		}
+
+		compl_params[i].event = odp_dma_compl_to_event(config->dma_config.compl_ev);
```

Comment:
If I understood the code correctly, it seems that multiple simultaneous async transfers may now share the same completion events (e.g. `odp_dma_perf -t 1 -g 1 -i 4 -m 1,1,1,1`). This is not allowed by the API.

Reply:
Yes, this is the case. Where is it stated that this is not allowed? (The command runs and completes fine.)

Reply:
This could perhaps be stated more clearly also in the DMA API, but the general ODP principle is that once an event is passed to scheduling, the caller loses access to the event handle. Only after receiving the event again, for example from `odp_schedule()`, can it be used again. Also, the same event can currently only exist once in scheduling/queues.

Reply:
OK, fixed in v4.